scripts/data_generation_offline.py

```diff
@@ -0,0 +1,362 @@
+        },
+        "format": {
+            "file_pattern": "data_{idx}.pt",
+            "fields": ["input_ids", "hidden_states", "loss_mask"],
+            "hidden_states_type": "List[torch.Tensor]",
+            "hidden_states_shape": "List of [seq_len, hidden_dim], one per layer",
+            "note": ("hidden_states is a list of tensors in order of layer_ids"),
+        },
+    }
+
+    config_path = os.path.join(output_dir, "data_config.json")
+    with open(config_path, "w") as f:
+        json.dump(config, f, indent=2)
+    log.info(f"Saved config to {config_path}")
+
+
+def generate_and_save_hidden_states(args, dataset):
+    """Generate hidden states and save each sample as a .pt file"""
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    start_file_idx = find_last_checkpoint(args.output_dir)
+    if start_file_idx > 0:
+        log.subsection(f"Resuming: {start_file_idx} files already exist")
+
+    num_samples = len(dataset)
+    start_sample_idx = start_file_idx - args.start_idx
+
+    if start_sample_idx >= num_samples:
+        log.info("All samples already processed!")
+        return 0
+
+    log.subsection("Initializing vLLM hidden states generator")
+    generator = VllmHiddenStatesGenerator(
+        model_path=args.target_model_path,
+        layer_ids=args.layer_ids,
+        max_model_len=args.max_model_len,
+        gpu_memory_utilization=args.gpu_memory_utilization,
+        tensor_parallel_size=args.tensor_parallel_size,
+    )
+
+    log.info(f"Processing {num_samples - start_sample_idx}/{num_samples} samples")
+    file_idx = start_file_idx
+
+    num_batches = (
+        num_samples - start_sample_idx + args.batch_size - 1
+    ) // args.batch_size
+
+    # Use ThreadPoolExecutor for async file I/O
+    max_io_workers = 4  # Number of parallel save operations
+
+    pbar = tqdm(
+        range(start_sample_idx, num_samples, args.batch_size),
+        desc="Generating hidden states",
+        total=num_batches,
+    )
+
+    with ThreadPoolExecutor(max_workers=max_io_workers) as thread_executor:
+        futures = []
+
+        for i in pbar:
+            batch_end = min(i + args.batch_size, num_samples)
+            batch = dataset[i:batch_end]
+            batch_input_ids = batch["input_ids"]
+            batch_loss_mask = batch["loss_mask"]
+
+            results = generator.generate(batch_input_ids)
+
+            # Submit save operations to thread pool (async I/O)
+            for j, result in enumerate(results):
+                result_cleaned = {
+                    "input_ids": result["input_ids"].clone(),
+                    "hidden_states": [h.clone() for h in result["hidden_states"]],
+                    "loss_mask": batch_loss_mask[j].clone(),
+                }
+                output_path = os.path.join(args.output_dir, f"data_{file_idx}.pt")
+                future = thread_executor.submit(
+                    save_sample_to_disk, result_cleaned, output_path
+                )
```

Comment:
Why not just call save_sample_to_disk in this thread? Otherwise you risk thread starvation and oom

Reply:
Also reduced overhead from thread launching.

Reply:
If generation is the bottleneck, you need disk writing to be coupled or else you'll oom
If disk writing is the bottleneck, decoupling the two step won't make total throughput any better.